File: src/agents/testRunner.ts
/**
 * Deterministic Test Runner (NO_LLM tier).
 *
 * The Test Runner must not consume an LLM request merely to run tests. This
 * executor detects the project's test framework, runs the relevant command
 * and returns a structured pass/fail SubagentResult — no model involved.
 *
 * Safety: only commands from the shared test-prefix allowlist are run; any
 * other command is reported as requiring approval, and commands never touch
 * source files. Output is bounded so a huge suite cannot blow up context.
 */
import { matchesPrefix, runShellCommand } from '../tools/runCommand.ts';
import { TEST_PREFIXES } from './capabilities.ts';
import type { ProjectInfo } from '../project/detect.ts';
import type { SubagentResult } from './types.ts';

const RUN_TIMEOUT_SEC = 300;
const MAX_SUMMARY_CHARS = 600;

/** Per-project-type fallback commands when no test command was detected. */
const TEST_COMMANDS_BY_TYPE: Record<string, string> = {
  node: 'npm test',
  python: 'python -m pytest',
  go: 'go test ./...',
  rust: 'cargo test',
  java: 'mvn test',
  ruby: 'bundle exec rspec',
};

function oneLiner(text: string): string {
  const flat = text.replace(/\s+/g, ' ').trim();
  return flat.length > MAX_SUMMARY_CHARS ? `${flat.slice(0, MAX_SUMMARY_CHARS - 1)}…` : flat;
}

export interface DeterministicTestRunnerOptions {
  projectRoot: string;
  project: ProjectInfo;
}

/** Run the project's tests without any model request. */
export async function runDeterministicTestRunner(opts: DeterministicTestRunnerOptions): Promise<SubagentResult> {
  const base: SubagentResult = {
    agent: 'test-runner',
    label: 'Test Runner',
    status: 'skipped',
    summary: '',
    files: [],
    changedFiles: [],
    findings: [],
    recommendations: [],
    iterations: 0,
    toolCalls: 0,
  };

  const command = opts.project.testCommand ?? TEST_COMMANDS_BY_TYPE[opts.project.type];
  if (!command) {
    return { ...base, status: 'skipped', summary: 'No test framework detected — nothing to run.' };
  }
  if (!matchesPrefix(command, TEST_PREFIXES)) {
    return {
      ...base,
      status: 'skipped',
      summary: `Tests need explicit approval ("${command}") — run them manually or approve first.`,
    };
  }

  const result = await runShellCommand(command, { cwd: opts.projectRoot, timeoutSec: RUN_TIMEOUT_SEC });
  const output = [result.stdout, result.stderr].filter(Boolean).join('\n').trim();

  if (result.timedOut) {
    return { ...base, status: 'failed', summary: `Tests timed out after ${RUN_TIMEOUT_SEC}s.` };
  }
  if (result.exitCode === 0) {
    return {
      ...base,
      status: 'completed',
      summary: output ? `Passed — ${oneLiner(output)}` : 'All tests passed.',
    };
  }
  const detail = oneLiner(output || 'No output captured.');
  const error = `Tests failed (exit ${result.exitCode ?? 'unknown'}): ${detail}`;
  return { ...base, status: 'failed', summary: error, error, findings: [`Tests failed with exit code ${result.exitCode ?? 'unknown'}.`] };
}

File: grace/agents/test_runner.py
"""Deterministic Test Runner (port of src/agents/testRunner.ts, NO_LLM tier)."""

from grace.agents.capabilities import TEST_PREFIXES
from grace.agents.types import SubagentResult
from grace.tools.run_command import matches_prefix, run_shell_command

RUN_TIMEOUT_SEC = 300
MAX_SUMMARY_CHARS = 600

# Per-project-type fallback commands when no test command was detected.
TEST_COMMANDS_BY_TYPE = {
    "node": "npm test",
    "python": "python -m pytest",
    "go": "go test ./...",
    "rust": "cargo test",
    "java": "mvn test",
    "ruby": "bundle exec rspec",
}


def _one_liner(text: str) -> str:
    flat = " ".join(text.split()).strip()
    return flat[:MAX_SUMMARY_CHARS - 1] + "…" if len(flat) > MAX_SUMMARY_CHARS else flat


def run_deterministic_test_runner(project_root: str, project) -> SubagentResult:
    base = SubagentResult(
        agent="test-runner",
        label="Test Runner",
        status="skipped",
        summary="",
    )

    command = project.testCommand or TEST_COMMANDS_BY_TYPE.get(project.type)
    if not command:
        return SubagentResult(agent="test-runner", label="Test Runner", status="skipped", summary="No test framework detected — nothing to run.")
    if not matches_prefix(command, TEST_PREFIXES):
        return SubagentResult(
            agent="test-runner",
            label="Test Runner",
            status="skipped",
            summary=f'Tests need explicit approval ("{command}") — run them manually or approve first.',
        )

    result = run_shell_command(command, cwd=project_root, timeout_sec=RUN_TIMEOUT_SEC)
    output = "\n".join(x for x in [result["stdout"], result["stderr"]] if x).strip()

    if result["timedOut"]:
        return SubagentResult(agent="test-runner", label="Test Runner", status="failed", summary=f"Tests timed out after {RUN_TIMEOUT_SEC}s.")
    if result["exitCode"] == 0:
        return SubagentResult(
            agent="test-runner",
            label="Test Runner",
            status="completed",
            summary=f"Passed — {_one_liner(output)}" if output else "All tests passed.",
        )
    detail = _one_liner(output or "No output captured.")
    error = f"Tests failed (exit {result['exitCode'] if result['exitCode'] is not None else 'unknown'}): {detail}"
    return SubagentResult(
        agent="test-runner",
        label="Test Runner",
        status="failed",
        summary=error,
        error=error,
        findings=[f"Tests failed with exit code {result['exitCode'] if result['exitCode'] is not None else 'unknown'}."],
    )
